### DISCUSSION TITLE
Issues encountered upgrading v0.17.3 -> v0.18.0

### DISCUSSION THREAD

**Discussion by mhratson** [vectordotdev/vector#10127] [Q&A] [answered]
## Issues encountered
They weren't mentioned in [upgrade guide](https://vector.dev/highlights/2021-11-18-0-18-0-upgrade-guide/) so  documenting it here for there record.

## `max_bytes` limit 4250000

Tests and validation haven't picked up

## Error
```
{"timestamp":"2021-11-19T20:32:53.271657Z","level":"ERROR","message":"Configuration error.","error":"Sink \"datadog\": provided `max_bytes` exceeds the maximum limit of 4250000","target":"vector::topology"}
```

### Config
Is no longer valid
```
  # max DD allows https://docs.datadoghq.com/api/latest/logs/#send-logs
  batch.max_bytes = 5242880
```

worked around by changing to `batch.max_bytes = 4250000` instead

## Improvements Encountered

Lower CPU usage
![image](https://user-images.githubusercontent.com/25061204/142699831-b27d11aa-c738-4f4f-87df-443b02bc4c85.png)

Lower Memory usage
![image](https://user-images.githubusercontent.com/25061204/142699713-91030ee0-d77c-401b-b708-8369c0602a5f.png)

**Comment by tobz** [comment, marked as answer]
@mhratson 👋🏻 

Sorry about that.  First and foremost, I opened an issue (#10128) for getting the upgrade guide fixed to properly call out the changes the batch size limits for the Datadog Logs sink, as well as other sinks.

For just a _small_ bit of context: the mismatch between the configurable maximum and what the Datadog documentation calls out is due to our internal mechanism for constructing batches.  Our imposed limit is lower so that we avoid generating payloads that are too big and would be requested by the Datadog API.

We actually have an issue open around this (#10020) because it can be confusing to users who think the batch size is equivalent to the payload size that Vector sends over the wire, when in reality, the batch size is an internal Vector concern around memory usage.  In practice, the payloads usually _are_ around the same size as the batches, but the current behavior can lead to undersized (most common) or oversized payloads (rare but possible).

**Reply by mhratson** [reply]
Thanks for the update @tobz.

>because it can be confusing to users who think the batch size is equivalent to the payload size that Vector sends over the wire

It definitely is. Would be nice to have internal setting derive its value from the one being configured by a user that may not have enough awareness about Vector internal peculiarities.

**Reply by tobz** [reply]
Right, that's the ultimate goal.
